### ISSUE TITLE
Documentation on custom null-ls formatters is incomplete/confusing

### ISSUE THREAD

**Issue by andrewferrier** [jose-elias-alvarez/null-ls.nvim#364]
### FAQ

- [X] I have checked the [FAQ](https://github.com/jose-elias-alvarez/null-ls.nvim#faq) and it didn't resolve my problem.

### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/null-ls.nvim/issues?q=is%3Aissue) and there are no issues with the same problem.

### Neovim Version

v0.5.1

### null-ls config

N/A

### Steps to reproduce

N/A

### Expected behavior

I've been trying to figure out how to use null_ls to specify some custom formatters. I have to say I did find the documentation a little confusing. The home README shows an example of how to use a generator to implement a custom diagnostics source (https://github.com/jose-elias-alvarez/null-ls.nvim#parsing-cli-program-output), but didn't contain any reference to how to implement a formatter. It took me a while to find https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/MAIN.md (this doesn't seem to be linked from the main README?), which gave me some clues, but even then this was a bit confusing. The section on formatting (https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/MAIN.md#formatting) appears to be incorrect; it doesn't show that the object should be returned from the `on_output` function, and in fact it appears that has a signature of `function(params, done)` which isn't shown in the code. In the end I had to copy-n-paste an implementation of `on_output` out of the `helpers.lua` file.

All in all, I think some more examples of how to create custom sources for formatting and other types apart from diagnostics would be really welcome.

I would be interested in helping you with this and sending in some pull request(s) if you'd like. One of the things I do find a bit odd is that there seems to be a distinction in `null_ls` between built-in sources and other sources, with slightly different functions for configuring these. I still don't really quite grok why this is.

I know I did open a bug on this repo before and you said you were trying to encourage people to add built-in sources. I agree, and I think that's a good goal, but it's important to make it easy for them to configure their own locally first to try them out before they send them in to you. I have a few I could send you but I'm not sure if I'm doing it the right way as I have nothing to crib from :)

Great project, `null_ls` works really well, and I would look forward to helping you if you're interested.

### Actual behavior

N/A

### Debug log

N/A

### Help

Yes

### Implementation help

Some guidance on the distinction between "built-in" sources and "non-built-in" sources; how are these different apart from the first obviously being shipped with `null_ls` and the latter not?

**Comment by jose-elias-alvarez** [comment]
I definitely agree that the documentation could use improvements. I've spent the past couple of weeks cleaning up issues and feature requests and won't have the bandwidth to work on anything but bug fixes for a while, but I'd be happy to review any PRs improving the situation.

To briefly explain the reasons for the distinction: First, built-in sources were not originally modularized, so they contain a now-unnecessary performance optimization whereby the generator is not actually set [until indexed](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/8d5082467246f46ee720ba7ef031e40add46643f/lua/null-ls/helpers.lua#L359-L363). The second big difference is the `with` method, which is meant to allow users to customize defaults. This is inherently unnecessary for custom sources, since the user who creates them can set whatever options fit their needs / environment.

To sum it up, `make_builtin` is only meant to be used for built-ins included in the null-ls repository. I agree that we could provide better examples of how to create sources without it, ideally one for each supported LSP method.

**Comment by andrewferrier** [comment]
OK, sounds good, and sounds like any doc improvements should focus on the non-`make_builtin` method. If I get some time, I'll send some PRs your way.

On 2021-11-19 at 11:22 -0800, Jose Alvarez ***@***.***> wrote:

> I definitely agree that the documentation could use improvements. I've spent the past couple of weeks cleaning up issues and feature requests and won't have the bandwidth to work on anything but bug fixes for a while, but I'd be happy to review any PRs improving the situation.
> 
> To briefly explain the reasons for the distinction: First, built-in sources were not originally modularized, so they contain a now-unnecessary performance optimization whereby the generator is not actually set [until indexed](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/8d5082467246f46ee720ba7ef031e40add46643f/lua/null-ls/helpers.lua#L359-L363). The second big difference is the `with` method, which is meant to allow users to customize defaults. This is inherently unnecessary for custom sources, since the user who creates them can set whatever options fit their needs / environment.
> 
> To sum it up, `make_builtin` is only meant to be used for built-ins included in the null-ls repository. I agree that we could provide better examples of how to create sources without it, ideally one for each supported LSP method. 
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/jose-elias-alvarez/null-ls.nvim/issues/364#issuecomment-974343773

-- 
Andrew Ferrier

+44 7968 147953
***@***.***
http://www.andrewferrier.com/

**Comment by matiascaniete** [comment]
Hi @andrewferrier, when I was trying to use shellcheck as a shell script formatter (which isn't a built-in so far) I also stumble upon with the same problem. So taking some clues from the docs on how to write a custom diagnostic, I found basic solution for implementing it:

```lua
local null_ls = require("null-ls")
local shellcheck_formatter = {
    method = null_ls.methods.FORMATTING,
    filetypes = { "sh" },
    generator = null_ls.formatter({
        command = "sh",
        args = { "-c", "shellcheck $0 --format=diff | patch $0 -o-", '$FILENAME' },
        to_stdin = true,
        from_stderr = true,
    }),
}
null_ls.register(shellcheck_formatter)
```

I hope this help.

**Comment by andrewferrier** [comment]
@matiascaniete did you mean to comment on this issue? Irrespective, shellcheck is a built in: https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/lua/null-ls/builtins/diagnostics/shellcheck.lua

**Comment by andrewferrier** [comment]
Scratch that, sorry, just realized you said 'formatter'. My bad!
